Tools/Cli-LoRa-Device-Provisioning/IoTDeviceHelper.cs

```diff
@@ -0,0 +1,1113 @@
+
+                if (!string.IsNullOrEmpty(opts.FCntResetCounter))
+                    twinProperties.Desired[TwinProperty.FCntResetCounter] = ValidationHelper.SetUIntTwinProperty(opts.FCntResetCounter);
+            }
+
+            // OTAA device specific properties
+            else
+            {
+                Console.WriteLine($"Creating OTAA device twin: {opts.DevEui}...");
+
+                twinProperties.Desired[TwinProperty.AppEUI] = ValidationHelper.SetStringTwinProperty(opts.AppEui);
+                twinProperties.Desired[TwinProperty.AppKey] = ValidationHelper.SetStringTwinProperty(opts.AppKey);
+            }
+
+            // Shared, non optional properties
+            twinProperties.Desired[TwinProperty.GatewayID] = ValidationHelper.SetStringTwinProperty(opts.GatewayId);
+            twinProperties.Desired[TwinProperty.SensorDecoder] = ValidationHelper.SetStringTwinProperty(opts.SensorDecoder);
+
+            // Shared, optional properties
+            if (!string.IsNullOrEmpty(opts.ClassType))
+                twinProperties.Desired[TwinProperty.ClassType] = ValidationHelper.SetStringTwinProperty(opts.ClassType);
+
+            if (!string.IsNullOrEmpty(opts.DownlinkEnabled))
+                twinProperties.Desired[TwinProperty.DownlinkEnabled] = ValidationHelper.SetBoolTwinProperty(opts.DownlinkEnabled);
+
+            if (!string.IsNullOrEmpty(opts.PreferredWindow))
+                twinProperties.Desired[TwinProperty.PreferredWindow] = ValidationHelper.SetUIntTwinProperty(opts.PreferredWindow);
+
+            if (!string.IsNullOrEmpty(opts.Deduplication))
+                twinProperties.Desired[TwinProperty.Deduplication] = ValidationHelper.SetStringTwinProperty(opts.Deduplication);
+
+            if (!string.IsNullOrEmpty(opts.Rx2DataRate))
+                twinProperties.Desired[TwinProperty.RX2DataRate] = ValidationHelper.SetStringTwinProperty(opts.Rx2DataRate);
+
+            if (!string.IsNullOrEmpty(opts.Rx1DrOffset))
+                twinProperties.Desired[TwinProperty.RX1DROffset] = ValidationHelper.SetUIntTwinProperty(opts.Rx1DrOffset);
+
+            if (!string.IsNullOrEmpty(opts.Supports32BitFCnt))
+                twinProperties.Desired[TwinProperty.Supports32BitFCnt] = ValidationHelper.SetBoolTwinProperty(opts.Supports32BitFCnt);
+
+            var twin = new Twin();
+            twin.Properties = twinProperties;
+
+            return twin;
+        }
+
+        public Twin UpdateDeviceTwin(Twin twin, Program.UpdateOptions opts)
+        {
+            Console.WriteLine();
+            Console.WriteLine($"Applying changes to device {opts.DevEui} twin...");
+
+            // ******************************************************************************************
+            // DevEui can NOT be updated!
+            // ******************************************************************************************
+
+            // ABP device properties
+            if (!string.IsNullOrEmpty(opts.AppSKey))
```

Comment:
Looks like there is quite a bit of shared code between Create and Update. You may want to add some helper methods to re-use the code.

Reply:
Discussed.